### DISCUSSION TITLE
Will curl consider supporting mbedtls 3?

### DISCUSSION THREAD

**Discussion by Benau** [curl/curl#7376] [Q&A] [answered]
Basically starting with newest mbedtls 3 release, md4.h no longer exists (which curl assume it always has), and some struct no longer exposing some public member (need some change to mbedtls_pk_parse_keyfile function used in lib/vtls/mbedtls.c)

**Comment by gilles-peskine-arm** [comment]
Mbed TLS developer here. Among the [changes in Mbed TLS 3.0](https://github.com/ARMmbed/mbedtls/releases/tag/v3.0.0), I think the ones that are most likely to affect curl are:

* Many obsolete cryptographic mechanisms and TLS protocol versions and options have been removed, including RC4, MD2, MD4, (D)TLS 1.0 and TLS 1.1.
* Fields in structures are now private by default. We're aware that we may have been overeager in some places, and we're likely to make a few fields public again in 3.x minor releases, or to add new accessor functions. If you have a good reason for a field to be public, please let us know on https://github.com/ARMmbed/mbedtls/issues/4682 or on the [mailing list](https://lists.trustedfirmware.org/mailman/listinfo/mbed-tls).

**Comment by bagder** [comment, marked as answer]
Sure, we'll be happy to support mbedTLS 3. Someone just needs to do the work...

**Comment by Benau** [comment]
https://github.com/Benau/curl/commit/a5a82e86b54c4754806d97e4f833446f485a3b8d

tested in android, @gilles-peskine-arm can you see if i have done something wrong?

basically seems that there is no way to get the keys structure without using MBEDTLS_PRIVATE, do you think those field curl using should remain unchanged in the near future?

also due to the removal of tls 1.0 1.1 some curl handling may be outdated here, not sure if I managed them correctly...
